src/dune_rules/jsoo/jsoo_rules.ml

```diff
@@ -48,85 +48,147 @@ module Config : sig
   val path : t -> string
   val of_string : string -> t
   val of_flags : string list -> t
-  val to_flags : t -> string list
+
+  (** [recent] should be true if jsoo version is 6.0 or higher. *)
+  val to_flags : recent:bool -> t -> string list
 end = struct
+  type effects_backend =
+    | Cps
+    | Double_translation
+
   type t =
     { js_string : bool option
-    ; effects : bool option
+    ; effects : effects_backend option
     ; toplevel : bool option
     }
 
   let default = { js_string = None; effects = None; toplevel = None }
   let bool_opt = [ None; Some true; Some false ]
+  let effects_opt = [ None; Some Cps; Some Double_translation ]
 
   let all =
     List.concat_map bool_opt ~f:(fun js_string ->
-      List.concat_map bool_opt ~f:(fun effects ->
+      List.concat_map effects_opt ~f:(fun effects ->
         List.concat_map bool_opt ~f:(fun toplevel -> [ { js_string; effects; toplevel } ])))
   ;;
 
-  let get t =
-    List.filter_map
-      [ "use-js-string", t.js_string; "effects", t.effects; "toplevel", t.toplevel ]
-      ~f:(fun (n, v) ->
-        match v with
-        | None -> None
-        | Some v -> Some (n, v))
-  ;;
-
   let set acc name v =
-    match name with
-    | "use-js-string" -> { acc with js_string = Some v }
-    | "effects" -> { acc with effects = Some v }
-    | "toplevel" -> { acc with toplevel = Some v }
+    match name, v with
+    | "use-js-string", `True -> { acc with js_string = Some true }
+    | "use-js-string", `False -> { acc with js_string = Some false }
+    | "effects", `Effects backend -> { acc with effects = Some backend }
+    | "effects", `False ->
+      (* [--disable effects] *)
+      { acc with effects = None }
+    | "effects", `True ->
+      (* [--enable effects], used alone, implies [--effects=cps] *)
+      (match acc.effects with
+       | None -> { acc with effects = Some Cps }
+       | Some _ -> acc)
+    | "toplevel", `True -> { acc with toplevel = Some true }
+    | "toplevel", `False -> { acc with toplevel = Some false }
```

Comment:
It would be simpler to have two `set` functions. One for `--enable/--disable` and one for `--effects`.